### PR TITLE
Save Optimizer Variables With Keras Fit API In Eager Mode

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- This PR allows the smdebug tensorflow hook to save optimizer variables in keras fit eager mode

- Removed an if statement in `set_optimizer_variables` that checked if the code was executing in eager mode and returned thereby not saving any optimizer variables.

- Introduced a new fn `_write_optimizer_variables` that writes  optimizer variables stored in collections when called

- Since optimizer variables are not available int the `_on_any_batch_begin` callback, 
`_write_optimizer_variables` is called in `_on_any_batch_end`

- Added an additional call to `self._prepare_non_layer_tensors()` to in `_on_any_batch_end` to populate addional collections like `ALL` with optimizer_variables before writing.

- Edited `_prepare_non_layer_tensors` to add tensor refs to custom collections

- I have also modified assert statements to validate the new tensor being saved.

The total time the test suite took to run on a Tensorflow 2.1 DLC was:

```
real	10m16.488s
user	5m42.804s
sys	6m1.492s
```

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Vikas-kum** [comment]
Change the description to say if it is eager, non eager mode ?

File: smdebug/tensorflow/base_hook.py [hunk 1]
```diff
@@ -401,6 +401,8 @@ def _get_collections_with_tensor(self, tf_tensor_name) -> Set["Collection"]:
         # matched with collections by regex
         if self.tape:
             return super()._get_collections_with_tensor(tf_tensor_name)
+        if tf_tensor_name not in self.tensor_to_collections and tf.executing_eagerly():
```
Comment (Vikas-kum) [inline smdebug/tensorflow/base_hook.py:404]: why is first check not in L402?

File: smdebug/tensorflow/base_hook.py [hunk 2]
```diff
@@ -489,12 +491,14 @@ def set_optimizer_variables(self, optimizer_variables):
         # From TF 2.2, executing_eagerly_outside_functions() can be used as
         # ops.executing_eagerly_outside_functions() or tf.compat.v1.executing_eagerly_outside_functions().
         # But in TF 2.1, only ops.executing_eagerly_outside_functions() is valid
-        if is_tf_version_2x() and ops.executing_eagerly_outside_functions():
```
Comment (Vikas-kum) [inline smdebug/tensorflow/base_hook.py:492]: why is this removed?

File: smdebug/tensorflow/base_hook.py [hunk 3]
```diff
@@ -489,12 +491,14 @@ def set_optimizer_variables(self, optimizer_variables):
...
-            return
         # since this is done for each variable at a time for keras, not checking if set already
         self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES).add_for_mode(
             optimizer_variables, ModeKeys.TRAIN
         )
+        if self.save_all is True:
```
Comment (Vikas-kum) [inline smdebug/tensorflow/base_hook.py:498]: earlier save_all method was not saving optimizer variables?

File: smdebug/tensorflow/keras.py [hunk 4]
```diff
@@ -323,9 +323,10 @@ def _prepare_non_layer_tensors(self):
                     self.tensor_to_collections[tensor_ref.name] = {coll}
                 elif coll not in self.tensor_to_collections[tensor_ref.name]:
                     self.tensor_to_collections[tensor_ref.name].add(coll)
+                if is_tf_version_2x() and tf.executing_eagerly():
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:326]: Not sure how is this change related to PR description?

File: smdebug/tensorflow/keras.py [hunk 5]
```diff
@@ -387,6 +388,9 @@ def _save_tensors_post_step(self, batch, logs):
         if is_tf_version_2x() and tf.executing_eagerly():
             for tensor_ref in self.tensor_refs_to_save_this_step:
                 tensor = tensor_ref.tf_obj
+                if tensor is None:
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:391]: is there a test case for this. Do we know when this can happen?

File: smdebug/tensorflow/base_hook.py [hunk 6]
```diff
@@ -489,12 +491,14 @@ def set_optimizer_variables(self, optimizer_variables):
         # From TF 2.2, executing_eagerly_outside_functions() can be used as
```
Comment (vandanavk) [inline smdebug/tensorflow/base_hook.py:491]: remove the comment ```# TF 2.x doesn't provide gradient/optimizer variable names and values by default.
        # Skipping set_gradients and set_optimizer_variables for Tf 2.x until there is
        # support to pass names and values from TF side.```

File: smdebug/tensorflow/base_hook.py (hunk 2 above)
Comment (vandanavk) [reply smdebug/tensorflow/base_hook.py:492]: this check was added to block set_optimizer_variables in eager mode until this PR which ensures capturing optimizer variable is possible

File: tests/tensorflow2/test_keras.py [hunk 7]
```diff
@@ -395,14 +395,15 @@ def test_keras_fit(out_dir, tf_eager_mode, saveall):
...
+            assert len(trial.tensor_names()) == (12 if is_tf_2_2() else 13)
         else:
             assert len(trial.tensor_names()) == 21
         assert len(trial.tensor_names(collection=CollectionKeys.BIASES)) == 2
         assert len(trial.tensor_names(collection=CollectionKeys.WEIGHTS)) == 2
     else:  # save the default losses and metrics
-        assert len(trial.tensor_names()) == (3 if is_tf_2_2() and tf_eager_mode else 4)
+        assert len(trial.tensor_names()) == (12 if is_tf_2_2() and tf_eager_mode else 13)
     assert len(trial.tensor_names(collection=CollectionKeys.LOSSES)) == 1
+    assert len(trial.tensor_names(collection=CollectionKeys.OPTIMIZER_VARIABLES)) == 5
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:406]: is this true in case of both, vanilla TF and custom TF?

Comment (vandanavk) [reply tests/tensorflow2/test_keras.py:406]: what's the code flow in case of vanilla TF? where is set_optimizer_variables called from?

File: tests/tensorflow2/test_keras.py [hunk 8]
```diff
@@ -521,8 +522,8 @@ def test_weights_collections(out_dir, tf_eager_mode):
 
     trial = smd.create_trial(path=out_dir)
     # can't save gradients in TF 2.x
-    assert len(trial.tensor_names()) == (5 if is_tf_2_2() and tf_eager_mode else 6)
-    assert len(trial.tensor_names(collection=CollectionKeys.BIASES)) == 0
+    assert len(trial.tensor_names()) == (7 if is_tf_2_2() and tf_eager_mode else 8)
+    assert len(trial.tensor_names(collection=CollectionKeys.BIASES)) == 2
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:526]: are biases saved when user selects only weights collection? is this expected behavior?

File: smdebug/tensorflow/base_hook.py [hunk 9]
```diff
@@ -489,12 +491,14 @@ def set_optimizer_variables(self, optimizer_variables):
...
         # ops.executing_eagerly_outside_functions() or tf.compat.v1.executing_eagerly_outside_functions().
         # But in TF 2.1, only ops.executing_eagerly_outside_functions() is valid
-        if is_tf_version_2x() and ops.executing_eagerly_outside_functions():
-            return
         # since this is done for each variable at a time for keras, not checking if set already
         self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES).add_for_mode(
             optimizer_variables, ModeKeys.TRAIN
         )
+        if self.save_all is True:
+            self.collection_manager.get(CollectionKeys.ALL).add_for_mode(
```
Comment (vandanavk) [inline smdebug/tensorflow/base_hook.py:499]: how has it been adding to "all" until now? why dint we need this check so far?

File: tests/tensorflow2/test_keras.py [hunk 10]
```diff
@@ -385,22 +385,21 @@ def test_gradtape_persistent(out_dir, saveall):
 def test_keras_fit(out_dir, tf_eager_mode, saveall):
     hook = smd.KerasHook(out_dir=out_dir, save_all=saveall)
     helper_keras_fit(
-        trial_dir=out_dir,
-        hook=hook,
-        eager=tf_eager_mode,
-        steps=["train", "eval", "predict", "train"],
```
Comment (NihalHarish) [inline tests/tensorflow2/test_keras.py:391]: Why has train been repeated here?

File: smdebug/tensorflow/keras.py [hunk 11]
```diff
@@ -508,8 +508,22 @@ def _on_any_batch_begin(self, batch, mode, logs=None):
...
+                    tensor = tensor_ref.tf_obj
+                    self._save_for_tensor(
+                        tensor_name=tensor.name,
+                        tensor_value=tensor.value(),
+                        check_before_write=False,
+                    )
+
         # Write the gradients of the past step if the writer is still available.
         if self.writer is not None or len(self.writer_map):
+            _write_optimizer_variables()
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:526]: can this be moved to save_tensors_post_Step() ?

File: smdebug/tensorflow/base_hook.py [hunk 12]
```diff
@@ -482,19 +484,21 @@ def set_optimizer_variables(self, optimizer_variables):
...
 
         # From TF 2.2, executing_eagerly_outside_functions() can be used as
         # ops.executing_eagerly_outside_functions() or tf.compat.v1.executing_eagerly_outside_functions().
         # But in TF 2.1, only ops.executing_eagerly_outside_functions() is valid
-        if is_tf_version_2x() and ops.executing_eagerly_outside_functions():
-            return
         # since this is done for each variable at a time for keras, not checking if set already
         self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES).add_for_mode(
-            optimizer_variables, ModeKeys.TRAIN
+            optimizer_variables, self.mode
```
Comment (vandanavk) [inline smdebug/tensorflow/base_hook.py:496]: why this change?

File: smdebug/tensorflow/base_hook.py [hunk 13]
```diff
@@ -482,19 +484,21 @@ def set_optimizer_variables(self, optimizer_variables):
         This method helps find the optimizer variables (such as momentum)
         :param optimizer_variables: list of tf.Variables/tf.Tensors/tf.MirroredVariables
         """
-        # TF 2.x doesn't provide gradient/optimizer variable names and values by default.
-        # Skipping set_gradients and set_optimizer_variables for Tf 2.x until there is
+        # TF 2.x provides only symbolic gradient variables that do not provide access to their values.
```
Comment (vandanavk) [inline smdebug/tensorflow/base_hook.py:487]: you can remove this comment from set_optimizer_variables

Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py:491]: Done;

Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py:491]: I have also moved _write_optimizers to save_post_step

File: smdebug/tensorflow/base_hook.py (hunk 12 above)
Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py:500]: Why is there a need to hardcode `ModeKeys.TRAIN`,  will this not lead to a case when the mode is actually EVAL or PREDICT but we save tensors with `ModeKeys.TRAIN`?

File: smdebug/tensorflow/base_hook.py (hunk 3 above)
Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py]: In Eager mode optimizer variables only become available in `_on_any_batch_end`. 
The population tensor refs to the ALL collection was happening only once in `_on_any_batch_begin`.

But since the ALL collection is not the only additional collection that might need to be populated with optimizer variables, I have added an additional `_prepare_non_layer_tensors()` call in `_on_any_batch_end`

File: smdebug/tensorflow/base_hook.py (hunk 9 above)
Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py]: See comment above.

File: smdebug/tensorflow/base_hook.py [hunk 14]
```diff
@@ -401,6 +401,12 @@ def _get_collections_with_tensor(self, tf_tensor_name) -> Set["Collection"]:
         # matched with collections by regex
         if self.tape:
             return super()._get_collections_with_tensor(tf_tensor_name)
+        if (
+            tf_tensor_name not in self.tensor_to_collections
+            and is_tf_version_2x()
+            and tf.executing_eagerly()
+        ):
```
Comment (vandanavk) [inline smdebug/tensorflow/base_hook.py:408]: can this if be merged with the one on L402?

File: tests/tensorflow2/test_keras.py [hunk 15]
```diff
@@ -395,11 +395,12 @@ def test_keras_fit(out_dir, tf_eager_mode, saveall):
     # can't save gradients in TF 2.x eager mode
     if saveall:  # save losses, metrics, weights, biases
         if tf_eager_mode:
-            assert len(trial.tensor_names()) == (7 if is_tf_2_2() else 8)
+            assert len(trial.tensor_names()) == (12 if is_tf_2_2() else 13)
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:398]: is this true in case of both, vanilla TF and custom TF?

File: smdebug/tensorflow/keras.py [hunk 16]
```diff
@@ -550,12 +550,28 @@ def on_test_batch_begin(self, batch, logs=None):
...
             return
 
         if not is_tf_version_2x() or (is_tf_version_2x() and not tf.executing_eagerly()):
             self._remove_fetches_and_callbacks(mode)
+
+        if is_tf_version_2x() and tf.executing_eagerly():
+            # Need to prepare non layer tensors again since
+            # some tensors only become available on  batch end
+            self._prepare_non_layer_tensors()
+            self._write_optimizer_variables()
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:573]: if these calls are made here, they must be executed for vanilla as well as custom TF. right?

custom TF also makes a call to set_optimizer_variables() ? does it try to set the same set of tensors or different? do we need an additional set_optimizer_variables() call if we get all the variables here?

File: tests/tensorflow2/test_keras.py (hunk 15 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:398]: Yes

File: smdebug/tensorflow/base_hook.py (hunk 14 above)
Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py:406]: Done.

File: smdebug/tensorflow/keras.py (hunk 16 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:573]: There is no difference in tensors saved when run on AWS TF v/s vanilla

File: smdebug/tensorflow/keras.py [hunk 17]
```diff
@@ -550,12 +550,28 @@ def on_test_batch_begin(self, batch, logs=None):
     def on_predict_batch_begin(self, batch, logs=None):
         self._on_any_batch_begin(batch, ModeKeys.PREDICT, logs=logs)
 
+    def _write_optimizer_variables(self):
+        optimizer_collections = self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES)
+        if optimizer_collections in self._get_collections_to_save_for_step():
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:555]: What if user provide regex for optimizer variables? Will it still be saved?
Can you please Try an example with : regex of optimizer variables 
and
A custom collection with regex of optimizer variables.

**Comment by Vikas-kum** [comment]
"Added an additional call to self._prepare_non_layer_tensors() to in _on_any_batch_end to populate addional collections like ALL with optimizer_variables before writing." 

QQ. Will optimizer variable be saved for eval batch/test batch also or just TRAIN batch ?

**Comment by NihalHarish** [comment]
> "Added an additional call to self._prepare_non_layer_tensors() to in _on_any_batch_end to populate addional collections like ALL with optimizer_variables before writing."
> 
> QQ. Will optimizer variable be saved for eval batch/test batch also or just TRAIN batch ?

Optimizers are added to collections with this line of code:

```
 self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES).add_for_mode(
           optimizer_variables, self.mode
```

And `_on_any_batch_end` is called by `on_train_batch_end`, `on_test_batch_end`, `on_predict_batch_end` and `on_eval_batch_end`, so they they will be saved in all modes

**Comment by Vikas-kum** [comment]
> > "Added an additional call to self._prepare_non_layer_tensors() to in _on_any_batch_end to populate addional collections like ALL with optimizer_variables before writing."
> > QQ. Will optimizer variable be saved for eval batch/test batch also or just TRAIN batch ?
> 
> Optimizers are added to collections with this line of code:
> 
> ```
>  self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES).add_for_mode(
>            optimizer_variables, self.mode
> ```
> 
> And `_on_any_batch_end` is called by `on_train_batch_end`, `on_test_batch_end`, `on_predict_batch_end` and `on_eval_batch_end`, so they they will be saved in all modes

Not sure what does it mean to save optimizer variables in test and predict and eval batch ?

File: smdebug/tensorflow/keras.py (hunk 17 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py]: I have added a test which creates custom collection with regex of optimizer variables.

`hook.get_collection("custom_optimizer_variables").include("Adam")`

**Comment by NihalHarish** [comment]
> > > "Added an additional call to self._prepare_non_layer_tensors() to in _on_any_batch_end to populate addional collections like ALL with optimizer_variables before writing."
> > > QQ. Will optimizer variable be saved for eval batch/test batch also or just TRAIN batch ?
> > 
> > 
> > Optimizers are added to collections with this line of code:
> > ```
> >  self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES).add_for_mode(
> >            optimizer_variables, self.mode
> > ```
> > 
> > 
> > And `_on_any_batch_end` is called by `on_train_batch_end`, `on_test_batch_end`, `on_predict_batch_end` and `on_eval_batch_end`, so they they will be saved in all modes
> 
> Not sure what does it mean to save optimizer variables in test and predict and eval batch ?

Optimizer variables are not present in EVAL and PREDICT

File: tests/tensorflow2/test_keras.py [hunk 18]
```diff
@@ -548,16 +550,18 @@ def test_include_collections(out_dir, tf_eager_mode):
         include_collections=include_collections,
         reduction_config=ReductionConfig(norms=ALLOWED_NORMS, reductions=ALLOWED_REDUCTIONS),
     )
+    hook.get_collection("custom_optimizer_variables").include("Adam")
```
Comment (Vikas-kum) [inline tests/tensorflow2/test_keras.py:553]: nice, can you write one more test case, 
so that mode save config is set in collection to save tensors in eval mode rather than TRAIN mode.
hook.get_collection("custom_optimizer_variables").save_config.mode_save_config 

Expectation should be num of tensors saved should be 0.

File: smdebug/tensorflow/keras.py [hunk 19]
```diff
@@ -317,12 +318,28 @@ def _prepare_layers(self, mode):
...
                 if tensor_ref.name not in self.tensor_to_collections:
-                    self.tensor_to_collections[tensor_ref.name] = {coll}
-                elif coll not in self.tensor_to_collections[tensor_ref.name]:
-                    self.tensor_to_collections[tensor_ref.name].add(coll)
+                    self.tensor_to_collections[tensor_ref.name] = {default_coll}
+                elif default_coll not in self.tensor_to_collections[tensor_ref.name]:
+                    self.tensor_to_collections[tensor_ref.name].add(default_coll)
+
+                # Add tensor to custom collections
+                for custom_coll in custom_collections:
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:338]: is this really reqd ? 

So what happens when, say you add only custom collection("optimizer_Variable").include("Adam.*") and this change is not there ? 

Don't we see Adam* tensors getting saved?

Tensors should be saved only once even if they are in different collection. Collection A can have tensorA, tensorB 
and Collection B can have tensorA and tensorB , 
but total number of tensors that should be saved is tensorA, tensorB .

File: tests/tensorflow2/test_keras.py (hunk 18 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:563]: Added an additional assert.

```
        assert (
            len(
                trial.tensor_names(
                    collection=CollectionKeys.OPTIMIZER_VARIABLES, mode=ModeKeys.EVAL
                )
            )
            == 0,
            "No Optimizer Variables Should be Saved in EVAL Mode",
        )
```

File: smdebug/tensorflow/keras.py (hunk 19 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:338]: The total number of tensors do not change with my changes.

Without my changes, the test fails for this particular case:

```
@pytest.mark.slow
def test_include_only_custom_collection(out_dir, tf_eager_mode):
    include_collections = ["custom_optimizer_variables"]
    save_config = SaveConfig(save_interval=3)
    hook = smd.KerasHook(
        out_dir,
        save_config=save_config,
        include_collections=include_collections,
        reduction_config=ReductionConfig(norms=ALLOWED_NORMS, reductions=ALLOWED_REDUCTIONS),
    )
    hook.get_collection("custom_optimizer_variables").include("Adam")
    helper_keras_fit(out_dir, hook=hook, steps=["train", "eval", "predict"], eager=tf_eager_mode)

    trial = smd.create_trial(path=out_dir)
    assert len(trial.tensor_names()) == (8 if is_tf_2_2() and tf_eager_mode else 9)
    assert len(trial.tensor_names(collection="custom_optimizer_variables")) == 5
```

Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:338]: I have added this test to the test suite as well
